CacheManager.StackExchange.Redis.SelfCleaning.Tests/SelfCleaningRedisIntegrationTests.cs

```diff
@@ -0,0 +1,247 @@
+                DateTime insertionTime = DateTime.Now;
+                keysWithInsertionTimes.Add((key, value, insertionTime));
+
+                Task.Delay(delayBetweenInsertionsSpan).Wait();
+            }
+
+            Task.Delay((int) (_timeToLive.TotalMilliseconds + TIME_TO_LIVE_MILLISECONDS_DIFFERENCE_THRESHOLD)).Wait();
+
+            // Assert
+            IDictionary<string, (CacheItemRemovedEventArgs Args, DateTime Time)> invocationsDictionary =
+                _onRemoveByHandleInvocations.ToDictionary(tuple => tuple.Args.Key, tuple => (tuple.Args, tuple.Time));
+
+            foreach ((string expectedKey, DummyModel expectedValue, DateTime insertionTime) in keysWithInsertionTimes)
+            {
+                CollectionAssert.Contains(invocationsDictionary.Keys, expectedKey);
+                OnRemoveByHandleAssertion(invocationsDictionary[expectedKey], expectedValue, insertionTime,
+                    _timeToLive);
+            }
+        }
+
+        [Test]
+        public void Put_AddKeyAndUpdateBeforeExpiry_TimeToLiveMeasuredFromLastUpdateTime()
+        {
+            // Arrange
+            var key = "key1";
+            var value = new DummyModel {Property = "property1"};
+            var updatedValue = new DummyModel {Property = "property2"};
+
+            TimeSpan delayBeforeUpdate = _timeToLive / 2;
+
+            // Act
+            _cache[key] = value;
+            DateTime insertionTime = DateTime.Now;
+
+            Task.Delay(delayBeforeUpdate).Wait();
+
+            _cache[key] = updatedValue;
+
+            Task.Delay((int) (_timeToLive.TotalMilliseconds + TIME_TO_LIVE_MILLISECONDS_DIFFERENCE_THRESHOLD)).Wait();
+
+            // Assert
+            OnRemoveByHandleAssertion(_onRemoveByHandleInvocations.Single(), updatedValue, insertionTime,
+                _timeToLive + delayBeforeUpdate);
+        }
+
+        [Test]
+        public void PutAndRemove_AddKeyAndRemoveBeforeExpiry_OnRemoveByHandleNotCalled()
+        {
+            // Arrange
+            var key = "key1";
+            var value = new DummyModel {Property = "property1"};
+
+            TimeSpan delayBeforeRemove = _timeToLive / 2;
+
+            // Act
+            _cache[key] = value;
+
+            Task.Delay(delayBeforeRemove).Wait();
+
+            _cache.Remove(key);
+
+            Task.Delay((int) (_timeToLive.TotalMilliseconds + TIME_TO_LIVE_MILLISECONDS_DIFFERENCE_THRESHOLD)).Wait();
+
+            // Assert
+            CollectionAssert.IsEmpty(_onRemoveByHandleInvocations);
+        }
+
+        private static void OnRemoveByHandleAssertion((CacheItemRemovedEventArgs, DateTime) invocation,
+            DummyModel expectedValue, DateTime insertionTime, TimeSpan expectedTimeAlive)
+        {
+            (CacheItemRemovedEventArgs args, DateTime removeTime) = invocation;
+
+            Assert.AreEqual(CacheItemRemovedReason.Expired, args.Reason);
+            Assert.AreEqual(expectedValue, args.Value);
```

Comment:
Maybe add this assert:
`Assert.IsInstanceOf<DummyModel>(args.Value);`

Im not sure if this is right, I jsut afraid of using AreEqual on type `object`...

Reply:
I've checked and internally it refers to `DummyModel.Equals`, so I say there's no need